File: examples/CPP/Geofencing/src/Geofencing.cpp
/*
 * Jimmy Paputto 2025
 */

#include <cstdio>

#include <jimmypaputto/GnssHat.hpp>


using namespace JimmyPaputto;

/*
 * Geofencing confidence levels map:
 * 0 - no confidence required
 * 1 - 68%
 * 2 - 95%
 * 3 - 99.7%
 * 4 - 99.99%
 * 5 - 99.9999%
 */

GnssConfig createDefaultConfig()
{
    const auto geofencing = GnssConfig::Geofencing {
        .geofences = std::vector<Geofence> {
            Geofence {
                .lat = 41.902205071091224,
                .lon = 12.4539203390548,
                .radius = 2005
            },
            Geofence {
                .lat = 52.257211745024186,
                .lon = 20.311759615806704,
                .radius = 1810
            }
        },
        .confidenceLevel = 3,
        .pioPinPolarity = EPioPinPolarity::LowMeansOutside
    };

    return GnssConfig {
        .measurementRate_Hz = 1,
        .dynamicModel = EDynamicModel::Stationary,
        .timepulsePinConfig = TimepulsePinConfig {
            .active = true,
            .fixedPulse = TimepulsePinConfig::Pulse { 1, 0.1 },
            .pulseWhenNoFix = std::nullopt,
            .polarity =
                ETimepulsePinPolarity::RisingEdgeAtTopOfSecond
        },
        .geofencing = geofencing
    };
}

void print(const Geofencing& geofencing)
{
    printf("Geofencing:\r\n");
    printf("  Configuration:\r\n");
    const auto confidenceLevel2str = [](const uint8_t l) -> const char*
    {
        switch (l)
        {
            case 0: return "No confidence required";
            case 1: return "68%";
            case 2: return "95%";
            case 3: return "99.7%";
            case 4: return "99.99%";
            case 5: return "99.9999%";
        }
        return "Fatal error";
    };
    printf("    Confidence level: %s\r\n",
        confidenceLevel2str(geofencing.cfg.confidenceLevel));
    printf("    Geofences:\r\n");
    for (const auto& geofence : geofencing.cfg.geofences)
    {
        printf("      Latitude: %f, Longitude: %f, Radius: %.2f m\r\n",
            geofence.lat, geofence.lon, geofence.radius);
    }

    printf("  Navigation:\r\n");

    const auto geofencingStatus2str =
        [](const EGeofencingStatus s) -> const char*
        {
            switch (s)
            {
                case EGeofencingStatus::NotAvalaible:
                    return "NotAvalaible";
                case EGeofencingStatus::Active:
                    return "Active";
            }
            return "Fatal error";
        };

    printf("    Geofencing status: %s\r\n",
        geofencingStatus2str(geofencing.nav.geofencingStatus));
    printf("    Number of geofences: %d\r\n", geofencing.nav.numberOfGeofences);

    const auto geofenceStatus2str =
        [](const EGeofenceStatus s) -> const char*
        {
            switch (s)
            {
                case EGeofenceStatus::Unknown: return "Unknown";
                case EGeofenceStatus::Inside: return "Inside";
                case EGeofenceStatus::Outside: return "Outside";
            }
            return "Fatal error";
        };

    printf("    Combined state: %s\r\n",
        geofenceStatus2str(geofencing.nav.combinedState));
    for (int8_t i = 0; i < geofencing.nav.numberOfGeofences; ++i)
    {
        printf("      Geofence no %d: %s\r\n",
            i+1, geofenceStatus2str(geofencing.nav.geofencesStatus[i]));
    }
}

auto main() -> int
{
    auto* ubxHat = IGnssHat::create();
    ubxHat->softResetUbloxSom_HotStart();
    const bool isStartupDone = ubxHat->start(createDefaultConfig());
    if (!isStartupDone)
    {
        printf("Startup failed, exit\r\n");
        return -1;
    }

    printf("Startup done, ublox configured\r\n");

    while (true)
    {
        const auto geofencing = ubxHat->waitAndGetFreshNavigation().geofencing;
        print(geofencing);
    }

    return 0;
}

File: examples/Python/geofencing.py
# Jimmy Paputto 2025

from jimmypaputto import gnsshat


def create_default_config() -> dict:
    """
    Geofencing confidence levels map:
    0 - no confidence required
    1 - 68%
    2 - 95%
    3 - 99.7%
    4 - 99.99%
    5 - 99.9999%
    """
    geofencing_config = {
        'confidence_level': 3,  # 99.7%
        'geofences': [
            {
                'lat': 41.902205071091224,
                'lon': 12.4539203390548,
                'radius': 2005
            },
            {
                'lat': 52.257211745024186,
                'lon': 20.311759615806704,
                'radius': 1810
            }
        ]
    }

    return {
        'measurement_rate_hz': 1,
        'dynamic_model': gnsshat.DynamicModel.STATIONARY,
        'timepulse_pin_config': {
            'active': True,
            'fixed_pulse': {
                'frequency': 1,
                'pulse_width': 0.1
            },
            'pulse_when_no_fix': None,
            'polarity': gnsshat.TimepulsePolarity.RISING_EDGE
        },
        'geofencing': geofencing_config
    }


def confidence_level_to_str(level: int) -> str:
    """Convert confidence level to human readable string"""
    confidence_map = {
        0: "No confidence required",
        1: "68%",
        2: "95%",
        3: "99.7%",
        4: "99.99%",
        5: "99.9999%"
    }
    return confidence_map.get(level, "Fatal error")


def geofencing_status_to_str(status: int) -> str:
    """Convert geofencing status to human readable string"""
    return gnsshat.GeofencingStatus(status).name


def geofence_status_to_str(status: int) -> str:
    """Convert geofence status to human readable string"""
    return gnsshat.GeofenceStatus(status).name


def print_geofencing(geofencing):
    """Print geofencing configuration and status"""
    print("Geofencing:")
    print("  Configuration:")

    confidence_str = confidence_level_to_str(geofencing.cfg.confidence_level)
    print(f"    Confidence level: {confidence_str}")

    print("    Geofences:")
    for geofence in geofencing.cfg.geofences:
        print(f"      Latitude: {geofence.lat:.6f}, "
              f"Longitude: {geofence.lon:.6f}, "
              f"Radius: {geofence.radius:.2f} m")

    print("  Navigation:")

    status_str = geofencing_status_to_str(geofencing.nav.status)
    print(f"    Geofencing status: {status_str}")
    print(f"    Number of geofences: {geofencing.nav.number_of_geofences}")
    
    combined_state_str = geofence_status_to_str(geofencing.nav.combined_state)
    print(f"    Combined state: {combined_state_str}")
    
    for i, fence_status in enumerate(geofencing.nav.geofences):
        status_str = geofence_status_to_str(fence_status)
        print(f"      Geofence no {i+1}: {status_str}")


def main():
    ubx_hat = gnsshat.GnssHat()
    ubx_hat.hard_reset_cold_start()

    is_startup_done = ubx_hat.start(create_default_config())
    if not is_startup_done:
        print("Startup failed, exit")
        return -1

    print("Startup done, ublox configured")

    try:
        while True:
            navigation = ubx_hat.wait_and_get_fresh_navigation()
            geofencing = navigation.geofencing
            print_geofencing(geofencing)
            print()  # Empty line for readability
            
    except KeyboardInterrupt:
        print("\nStopping geofencing monitoring...")
    except Exception as e:
        print(f"Error: {e}")
        return -1
    
    print("Program finished")
    return 0


if __name__ == "__main__":
    exit(main())
